### PR TITLE
AirfRANS: lr=8e-4+T_max=10 — continue LR upper-bound search

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

lr=7e-4 produced an earlier phase transition (epoch 35 vs 38-40 for lower LRs) and the best result yet (0.01841 val, PR #2646). Testing lr=8e-4 to see if pushing LR higher continues to deepen the basin. Note: lr=1e-3 was previously tested (senku #2615) but at a different baseline config — it's worth checking the boundary between 7e-4 and 1e-3 to understand the LR sensitivity and whether there is more headroom.

Key question: Does the higher LR continue to accelerate the phase transition and find a deeper basin, or does it destabilize training?

## Baseline to Beat

- **Primary metric:** `val_primary/surface_mse`
- **Current best:** 0.01841 val (PR #2646, lr=7e-4+T_max=10, W&B: 3pbxocca)
- **Reproduce baseline:** `cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 10 --no-use-ema --enable-fourier --epochs 999`

## Instructions

Run the following command. Training will take ~30 min (SENPAI timeout). Capture the best `val_primary/surface_mse` across all epochs (use W&B — the best epoch may not be the final epoch due to cosine LR cycling).

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 8e-4 --cosine-t-max 10 \
  --no-use-ema --enable-fourier --epochs 999 \
  --agent edward --wandb-name "edward/airfrans-lr8e4-tmax10" \
  --wandb-group "airfrans-lr-sweep-v2"
```

**W&B group:** `airfrans-lr-sweep-v2`

## What to Report

- Best `val_primary/surface_mse` (check all epochs in W&B — it will be at a cosine trough)
- Which epoch the best appeared at (to check phase transition timing vs lr=7e-4's epoch 35)
- Whether training was stable (no divergence/explosion at cosine peaks)
- W&B run ID

## Context: LR Sweep History

| LR    | Best val_primary/surface_mse | Phase transition epoch | PR    |
|-------|------------------------------|------------------------|-------|
| 3e-4  | 0.0197                       | ~38                    | #2614 |
| 5e-4  | 0.0207                       | ~40                    | #2617 |
| 7e-4  | 0.01841                      | 35                     | #2646 |
| 8e-4  | ?                            | ?                      | this  |

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### lr=8e-4 + T_max=10 (AirfRANS full)

| Metric | Baseline (lr=7e-4) | lr=8e-4 | Delta |
|--------|-------------------|---------|-------|
| **val_primary/surface_mse** | **0.01841** | **0.03833** | +108% (worse) |
| full_val/surface_mse_p | 0.0735 | 0.1531 | +108% |
| full_val/volume_mse | 0.1331 | 0.1467 | +10% |
| test_primary/surface_mse | — | 0.1591 | (at LR peak, misleading) |

- **Best val_primary/surface_mse:** 0.03833 at **epoch 29** (cosine trough)
- **Second best:** 0.03894 at epoch 39
- **Phase transition onset:** ~epoch 18-20 (vs epoch 35 for lr=7e-4 — **15 epochs earlier**)
- **Training stability:** Highly volatile. Peak-to-trough oscillations ranged 0.038–0.574 (vs much smaller swings at lr=7e-4). The LR peaks cause the model to be kicked far from the basin.
- **Epochs:** 41 (30-min timeout)
- **W&B run:** djl4y4o5
- **Peak memory:** ~6-8 GB (standard AirfRANS Fourier config)

### Command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 8e-4 --cosine-t-max 10 \
  --no-use-ema --enable-fourier --epochs 999 \
  --agent edward --wandb-name "edward/airfrans-lr8e4-tmax10" \
  --wandb-group "airfrans-lr-sweep-v2"
```

### What happened

**lr=8e-4 did NOT beat the baseline.** The higher LR accelerated the phase transition onset by ~15 epochs (epoch 18-20 vs 35), confirming the trend that higher LR triggers earlier transitions. However, it also dramatically destabilized training:

1. **Violent oscillations:** The cosine LR peaks at 8e-4 kick the model too far from the basin. Spikes to 0.57 (epoch 22) mean the model repeatedly loses progress gained at troughs.
2. **Shallower basin:** The deepest trough (0.038 at epoch 29) was 2x worse than lr=7e-4's best (0.01841 at epoch 35). The higher LR explores broadly but can't settle into the deepest minimum.
3. **No recovery after epoch 29:** Troughs at epochs 36-39 (0.078, 0.078, 0.039) never improved beyond the epoch 29 best, suggesting the training had found its limit at this LR.

### Trough envelope (cosine minima)
| Epoch | val_primary/surface_mse |
|-------|------------------------|
| 6     | 0.2399                 |
| 9     | 0.2216                 |
| 12    | 0.1828                 |
| 18    | 0.1146                 |
| 20    | 0.0739                 |
| 25    | 0.0916                 |
| 28    | 0.0578                 |
| 29    | **0.0383**             |
| 36    | 0.0780                 |
| 37    | 0.0776                 |
| 39    | 0.0389                 |

### Interpretation: LR sensitivity landscape

Updated sweep table:

| LR    | Best val_primary/surface_mse | Phase transition epoch | Stability | PR    |
|-------|------------------------------|------------------------|-----------|-------|
| 3e-4  | 0.0197                       | ~38                    | Stable    | #2614 |
| 5e-4  | 0.0207                       | ~40                    | Stable    | #2617 |
| 7e-4  | **0.01841**                  | 35                     | Moderate  | #2646 |
| 8e-4  | 0.03833                      | ~20                    | Volatile  | this  |

**lr=7e-4 appears to be near the optimal boundary.** Going higher accelerates the transition but introduces too much instability for the cosine cycling to find the deepest basin. The sweet spot is where the LR is high enough to trigger the phase transition earlier but not so high that peak oscillations destroy the basin.

### Suggested follow-ups

1. **lr=7.5e-4** — probe the boundary between 7e-4 (stable, 0.01841) and 8e-4 (volatile, 0.038) more finely
2. **lr=8e-4 with T_max=15 or 20** — longer cosine cycles would reduce the frequency of destructive LR peaks, potentially letting 8e-4 settle into deeper basins
3. **lr=8e-4 with gradient clipping** — could tame the oscillations while keeping the fast phase transition

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2674 because ADVISOR: Closing — AirfRANS lr=8e-4 reached 0.03833 (108% worse than 0.01841). lr=8e-4 triggers phase transition earlier but with catastrophic volatility at peaks. Confirms lr=7e-4 is at or near the optimal boundary. DEAD END..
